File: test-dnd/TestDnD.java
import javax.swing.*;
import java.awt.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.dnd.*;

public class TestDnD {
    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> {
            JFrame frame = new JFrame("Java DnD Test");
            frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
            frame.setSize(400, 300);
            frame.setLocationRelativeTo(null);

            JLabel label = new JLabel("Drop a file here", SwingConstants.CENTER);
            label.setFont(new Font("Arial", Font.BOLD, 20));
            frame.getContentPane().add(label, BorderLayout.CENTER);

            DropTarget dt = new DropTarget(frame.getContentPane(), DnDConstants.ACTION_COPY,
                new DropTargetListener() {
                    public void dragEnter(DropTargetDragEvent e) {
                        System.out.println("dragEnter!");
                        e.acceptDrag(DnDConstants.ACTION_COPY);
                        label.setText("DRAGGING!");
                    }
                    public void dragOver(DropTargetDragEvent e) {
                        e.acceptDrag(DnDConstants.ACTION_COPY);
                    }
                    public void dropActionChanged(DropTargetDragEvent e) {}
                    public void dragExit(DropTargetEvent e) {
                        System.out.println("dragExit");
                        label.setText("Drop a file here");
                    }
                    public void drop(DropTargetDropEvent e) {
                        e.acceptDrop(DnDConstants.ACTION_COPY);
                        try {
                            java.util.List<?> files = (java.util.List<?>)
                                e.getTransferable().getTransferData(DataFlavor.javaFileListFlavor);
                            System.out.println("Dropped " + files.size() + " files!");
                            label.setText("Dropped " + files.size() + " files!");
                        } catch (Exception ex) {
                            ex.printStackTrace();
                        }
                        e.dropComplete(true);
                    }
                }, true);

            frame.setVisible(true);
            System.out.println("Ready - drag a file onto this window");
        });
    }
}

File: test-dnd/TestDnD.kt
import java.awt.BorderLayout
import java.awt.datatransfer.DataFlavor
import java.awt.dnd.DnDConstants
import java.awt.dnd.DropTarget
import java.awt.dnd.DropTargetDragEvent
import java.awt.dnd.DropTargetDropEvent
import java.awt.dnd.DropTargetEvent
import java.awt.dnd.DropTargetListener
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.SwingConstants

fun main() {
    val frame = JFrame("DnD Test")
    frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE
    frame.setSize(400, 300)

    val label = JLabel("Drop files here", SwingConstants.CENTER)
    frame.contentPane.add(label, BorderLayout.CENTER)

    val listener = object : DropTargetListener {
        override fun dragEnter(e: DropTargetDragEvent) {
            println("SWING: dragEnter")
            e.acceptDrag(DnDConstants.ACTION_COPY)
            label.text = "Dragging over..."
        }
        override fun dragOver(e: DropTargetDragEvent) {
            e.acceptDrag(DnDConstants.ACTION_COPY)
        }
        override fun dropActionChanged(e: DropTargetDragEvent) {}
        override fun dragExit(e: DropTargetEvent) {
            println("SWING: dragExit")
            label.text = "Drop files here"
        }
        override fun drop(e: DropTargetDropEvent) {
            e.acceptDrop(DnDConstants.ACTION_COPY)
            val t = e.transferable
            if (t.isDataFlavorSupported(DataFlavor.javaFileListFlavor)) {
                val files = t.getTransferData(DataFlavor.javaFileListFlavor) as List<*>
                println("SWING: dropped ${files.size} files")
                label.text = "Dropped ${files.size} files!"
            }
            e.dropComplete(true)
        }
    }

    frame.contentPane.dropTarget = DropTarget(frame.contentPane, DnDConstants.ACTION_COPY, listener, true)
    frame.isVisible = true
    println("Swing DnD test window ready")
}
